src/python/pants/backend/go/pkg.py

```diff
@@ -0,0 +1,256 @@
+            ]
+        ),
+    )
+
+    # Note: The `go` tool requires GOPATH to be an absolute path which can only be resolved from within the
+    # execution sandbox. Thus, this code uses a bash script to be able to resolve that path.
+    # TODO: Merge all duplicate versions of this script into a single script and add an invoke rule that will
+    # insert the desired `go` command into the boilerplate portions.
+    analyze_script_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "analyze.sh",
+                    textwrap.dedent(
+                        f"""\
+                export GOROOT="$(/bin/pwd)/go"
+                export GOPATH="$(/bin/pwd)/gopath"
+                export GOCACHE="$(/bin/pwd)/cache"
+                /bin/mkdir -p "$GOPATH" "$GOCACHE"
+                cd {resolved_go_module.address.spec_path}
+                exec "${{GOROOT}}/bin/go" list -json ./{spec_subpath}
+                """
+                    ).encode("utf-8"),
+                )
+            ]
+        ),
+    )
+
+    input_root_digest = await Get(
+        Digest,
+        MergeDigests([sources.snapshot.digest, downloaded_goroot.digest, analyze_script_digest]),
+    )
+
+    process = Process(
+        argv=[bash.path, "./analyze.sh"],
+        input_digest=input_root_digest,
+        description="Resolve go_package metadata.",
+        level=LogLevel.DEBUG,
+    )
+
+    result = await Get(ProcessResult, Process, process)
+
+    metadata = json.loads(result.stdout)
+
+    # TODO: Raise an exception on errors. They are only emitted as warnings for now because the `go` tool is
+    # flagging missing first-party code as a dependency error. But we want dependency inference and won't know
+    # what the dependency actually is unless we first resolve the package with that dependency. So circular
+    # reasoning. We may need to hydrate the sources for all go_package targets that share a `go_module`.
+    if metadata.get("Incomplete"):
+        error_dict = metadata.get("Error", {})
+        if error_dict:
+            error_str = error_to_string(error_dict)
+            _logger.warning(
+                f"Error while resolving Go package at address {request.address}: {error_str}"
+            )
+        # TODO: Check DepsErrors key as well.
+
+    # Raise an exception if any unsupported source file keys are present in the metadata.
+    for key in (
+        "CompiledGoFiles",
+        "CFiles",
+        "CXXFiles",
+        "MFiles",
+        "HFiles",
+        "FFiles",
+        "SFiles",
+        "SwigFiles",
+        "SwigCXXFiles",
+        "SysoFiles",
+    ):
+        files = metadata.get(key, [])
+        if files:
+            raise ValueError(
+                f"The go_package at address {request.address} contains the following unsupported source files "
+                f"that were detected under the key '{key}': {', '.join(files)}."
+            )
```

Comment:
What's the plan for this eventually? Usually we try w/ dep inference to not fail your build. The worst that happens is dep inference no-ops, and you have to explicitly add the dep.

This is fine while experimenting, but I suspect we don't want to error like this in the future.

Reply:
All of these keys are for non-Go files (e.g., Objective-C) that somehow the `go` tool is able to at least list in the metadata (and maybe deal with them?). I have "cgo_sources" (i.e., go interfacing with C) as a analyzed field but the rest of these are unlikely to ever be supported by Pants, if they are even supported by the `go` tool.

Reply:
And so the plan for these is to not support them and to flag them when they are present so the user knows we don't support them.